diesel_cli/src/main.rs

```diff
@@ -73,6 +76,48 @@ fn run_migration_command(matches: &ArgMatches) {
             let database_url = database::database_url(matches);
             call_with_conn!(database_url, redo_latest_migration);
         }
+        ("list", Some(args)) => {
+            use std::cmp::Ordering;
+
+            let database_url = database::database_url(matches);
+            let dir = migrations_dir(args);
+            let migrations = call_with_conn!(database_url, migrations::mark_migrations_in_directory(&dir, &mut stdout()))
+                .unwrap_or_else(handle_error);
+
+            // Since our migrations came from `mark_migrations_in_directory` they should all have valid file names.
+            let mut sorted = migrations.into_iter().map(|(path_buf, applied)| {
+                let file_name = path_buf.clone()
+                    .and_then(|p| p.file_name()
+                              .and_then(|os_str| os_str.to_str()
+                                        .map(|s| s.to_string())))
+                    .expect(&format!("Error getting file-name from {:?}", path_buf));
```

Comment:
This looks more complicated than it needs to be. You can make the chaining a bit easier, e.g. by

- using UFCS (`str::to_string` instead of `|s| s.to_string` which tells us what type/trait we are dealing with),
- and/or by using the "`Option::iter` trick" that treats the Option as an iterator (yielding exactly one value) so you can call `flat_map` on it (which consumes stuff that can be turned into an Iterator – including Option and Result!):

    ```rust
    let file_name = path_buf.iter()
        .flat_map(|p| p.file_name())
        .flat_map(::std::ffi::OsStr::to_str)
        .map(str::to_string)
        .next()
        .expect(&format!("Error getting file-name from {:?}", path_buf));
    ```

    Don't overdo this :)

Once thing I missed at first was that `path_buf` is actually an `Option<PathBuf>` and the expect at the end also panics when this is `None`. I'd make that explicit, which allows us to go back to regular chaining without `.iter()`:

```rust
let path_buf = path_buf.expect(&format!("Migration without valid file name"));
let file_name = path_buf.file_name()
    .and_then(::std::ffi::OsStr::to_str)
    .map(str::to_string)
    .expect(&format!("Error getting file name from {:?}", path_buf));
```